### DISCUSSION TITLE
Voice dir creation

### DISCUSSION THREAD

**Discussion by anycolorsmith** [probonopd/MiniDexed#174] [Q&A] [answered]
![WhatsApp Image 2022-04-24 at 14 24 01](https://user-images.githubusercontent.com/104384882/165303676-c6a9ead3-6ec8-41dd-b2a8-f3ea407dad9e.jpeg)
Hello, dear frends and creators! How i can made right directory for bank and voices? On HDMI screen shows message "sysex\voice directory not found"  made his in root, how showed in Floyd video. And how i can create voice dir and sort by name\author\instrument? 
This project very inspired me! In the past I have owned DX7 and EX5.

**Comment by eLuSiVeMiTe** [comment]
If you have banks in that folder and can't access them it's probably due to file names. Plug it into a hdmi monitor and will probably say file name no recognised etc. 

I numbered my banks and it found them.


Some clarification on what naming conventions we can use would be helpful. 

Also can folders be used within the voice folder?

**Comment by craigyjp** [comment, marked as answer]
It's called sysex/voice and the file names have to be numbered as follows

000000_filename.syx
000001_filename.syx

Etc

**Reply by eLuSiVeMiTe** [reply]
Cheers. 

Do you know if it will read within subfolders. Not had a chance to test yet

**Reply by craigyjp** [reply]
I don't know if it will, it doesn't make sense, as how do you address a bank within a bank with a midi bank change.

**Reply by eLuSiVeMiTe** [reply]
> I don't know if it will, it doesn't make sense, as how do you address a bank within a bank with a midi bank change.

Fair point.  

Selectable voice folders maybe.  

Non issue really but I'm on the habit of organising my stuff for easy searching.

**Reply by anycolorsmith** [reply]
> It's called sysex/voices and the file names have to be numbered as follows
> 
> 000000_filename.syx 000001_filename.syx
> 
> Etc

Not 'voices' but 'voice' Now works correctly and systen found and load bank, I try only one bank in folder.

**Reply by craigyjp** [reply]
Good, at least it is working for you now.

**Reply by anycolorsmith** [reply]
> Good, at least it is working for you now.

I try many banks ( two) and banks loaded sucesfully. I don't have yet LCD screen and encoder for check.

**Reply by probonopd** [reply]
I have added this information to https://github.com/probonopd/MiniDexed/wiki/Files.

**Reply by Andy2No** [reply]
@probonopd There seems to be a typo in https://github.com/probonopd/MiniDexed/wiki/Files#voice-data-syx-files :

```
The files have to be numbered and named in this way:

sysex/voice/000001_filename.syx
sysex/voice/000002filename.syx
...
```
Shouldn't there be an underscore in the second filename too?

**Reply by Banana71** [reply]
Hi Andy, thanks for your information. It's done.

**Comment by craigyjp** [comment]
I have 60+ banks loaded no problems

**Comment by aspat** [comment]
Is there any command line or script to copy various .syx-files from folder tree to proper Minidexed naming format?

**Reply by probonopd** [reply]
Currently there is only https://github.com/probonopd/MiniDexed/blob/main/getsysex.sh

**Reply by craigyjp** [reply]
I used this to work out the directory structure and naming convention used to create the banks. It was helpful, so I'm glad you have put this information into the wiki.

**Reply by probonopd** [reply]
This one gives you 1450 banks of voices, but unfortunately many of them don't work (yet? Probably because actually, after 000127... needs to come 001000, 001002,...). With so many banks, MiniDexed takes a while to boot. Use HDMI to see what is going on. I need to rewrite this properly in Python and make it also preserve the filenames.

```
#!/bin/bash

wget http://bobbyblues.recup.ch/yamaha_dx7/patches/DX7_AllTheWeb.zip
unzip DX7_AllTheWeb.zip

mkdir -p sysex/voice/
find DX7_AllTheWeb/ -type f -name '*.syx' -exec mv {} sysex/voice/ \;

a=1
for i in sysex/voice/*.syx; do
  new=$(printf "sysex/voice/%06d.syx" "$a") #06 pad to length of 6
  mv -i -- "$i" "$new"
  let a=a+1
done
```

**Reply by anycolorsmith** [reply]
> 

This script work with synth runned  on Raspberry? Just need keyboard?

**Reply by probonopd** [reply]
This script needs to be run on your computer. Not on the Raspberry Pi running MiniDexed. There is no bash in MiniDexed.

**Comment by craigyjp** [comment]
I created 128 banks on my SD card using the 000000_filename.syx format to 000127_filename.syx

I then created some banks 001000_filename.syx to 001014_filename.syx

I cannot see or access the second bank, maybe I'm doing something wrong or its not supported yet.

**Reply by aspat** [reply]
I can confirm that only the first 127 banks are selectable (Raspi 2B). Over 3500 syx-files in the SD-card...
Also there are many patches with weird/blank names and/or no sound.

**Reply by probonopd** [reply]
Indeed. Maybe I was wrong with my numbering theory ("after 000127... needs to come 001000, 001002,...").

**Comment by dhstsw** [comment]
Yes, i see, there's a limit to 128 soundbanks :-/

**Reply by probonopd** [reply]
is there? Why? Can we remove the limit?

**Comment by Andy2No** [comment]
I'm not sure if this is relevant to the limit on the number of sound banks that can be selected, or if that's still a problem, but I've found MiniDexed ignores Bank Select MSB, unless it's followed by Bank Select LSB.

I can send just Bank Select LSB followed by Program Change, and that works as expected, but if I send Bank Select MSB then Program Change, it stays in the same bank.  To switch to another one, the only way via MIDI seems to be to send the whole sequence:

Bank Select MSB, Bank Select LSB, Program Change.

I'm still trying this without buttons or a display, so I'm checking by sending two program changes, going back to where I was with the second one.  To have it react how I expected, to Bank Select MSB, I can follow it with two Bank Select LSB messages, with the second one selecting the same value I was on before, then send the Program Change.

I'm not sure if this was the intended behaviour, but I assumed I'd just be able to send either Bank Select LSB or Bank Select MSB, then a Program Change - I expected Bank Select LSB to be remembered as state information when I send just Bank Select MSB, so I wouldn't need to send both.   That doesn't seem to be what's happening though.

FWIW, I'm using the latest build from earlier today, which was MiniDexed_2023-07-09-571816f.  I haven't made any changes to the .ini files, I just added the curated library of patches/banks (https://github.com/probonopd/MiniDexed/discussions/456#discussioncomment-5710503), and pwm sound.  I'm using a blue plastic M-Audio Uno, which seems usable apart from notes sticking every couple of minutes.  I've tested bank select by using Virtual MIDI Keyboard by Wouter van Beek, with the two virtual user knobs set up as Bank Select MSB and Bank Select LSB - verified using MIDI-OX.

I'm using a newly acquired Pi 4B, which is getting a lot hotter than the Pi 3B v1.2 does, while apparently doing the same thing, in the same way.

**Reply by probonopd** [reply]
> I can send just Bank Select LSB followed by Program Change, and that works as expected, but if I send Bank Select MSB then Program Change, it stays in the same bank. To switch to another one, the only way via MIDI seems to be to send the whole sequence:
>
> Bank Select MSB, Bank Select LSB, Program Change.

[MIDI 1.0 Detailed Specification 4.2](https://ia801808.us.archive.org/6/items/complete_midi_96-1-3/complete_midi_96-1-3.pdf) says on page 13:

> The transmitter must transmit the MSB and LSB as a pair, and the Program Change must be sent
immediately after the Bank Select pair. If their is any delay between these messages and they are
passed through a merging device (which may insert another message) the message may be interpreted
incorrectly. 

So MiniDexed behaves correctly.

**Reply by probonopd** [reply]
> I'm using a newly acquired Pi 4B, which is getting a lot hotter than the Pi 3B v1.2 does

I don't think at the moment you are getting any benefits from using the 4B over the 3B.

**Reply by pik33** [reply]
4B has better audio quality on builtin audio jack, as the base PLL frequency is 750 MHz instead of 500. This gives at least 6dB better SNR.
As MiniDexed runs on 3B, 4B may be underclocked and undervoltaged. That should make it run cooler.

**Comment by Andy2No** [comment]
@probonopd I see.  That's going to make it harder to use from a keyboard, unfortunately.  Some synths just use one or the other - either MSB or LSB, so they don't follow that part of the MIDI spec by requiring both to be sent.  I wasn't aware that was the rule.

Yes, I think the 3B might be the better choice.  I found a comparison chart which shows the 4B uses considerably more power just idling.  The 3A+ uses similar current/power to the 3B, and the 3B+ is somewhere between the 3B/3A+ and the 4B.

<img width="1031" alt="Raspberry-Pi-4B-Power-Usage-chart-annotated" src="https://github.com/probonopd/MiniDexed/assets/8021302/ea7b8dbb-7af5-41f2-ad3c-f9ff2a91f00d">

**Reply by probonopd** [reply]
> Some synths just use one or the other - either MSB or LSB, so they don't follow that part of the MIDI spec by requiring both to be sent.

Do you have an example? That'd clearly violate the spec.

**Reply by poetaster** [reply]
The interfaces on the AKAI MPK 25 also CAN send MSB and LSB separately when using program change with bank selection. I've been trying to test with select PROG, MSB, LSB and then send, but I'm too confused to evaluate the results :)  What does work consistantly is using PROG and LSB together (or just switching LSB or program through 127). 

I have three folders in sysex/voices folder (oo1blah, 002foo, 003bar) and can't keep track. What would be a could test strategy? Or should I simply be naming the files with bank info? The docs are a bit confusing. And I'm a bit confused :)

**Reply by Banana71** [reply]
Hi poetaster,

> I have three folders in sysex/voices folder (oo1blah, 002foo, 003bar) and can't keep track. What would be a could test strategy? Or should I simply be naming the files with bank info? The docs are a bit confusing. And I'm a bit confused :)

The file names of the syx files are crucial.
Look at the following folder structure with the sysex files: MiniDexed-sysexFinal.zip https://github.com/probonopd/MiniDexed/discussions/456#discussioncomment-6406179

Peter 
🍌

**Comment by Andy2No** [comment]
The Korg Minilogue XD does it.

```
    KORG minilogue xd Implementation         Revision 1.01 (2020.2.10)

 1.TRANSMITTED DATA

1-1 CHANNEL MESSAGES            [H]:Hex,  [D]:Decimal


| Status | Second  |   Third     |           Description                               |
| [Hex]  | [H] [D] | [H]   [D]   |                                                     |
+--------+---------+-------------+-----------------------------------------------------+
...
|   Bn   | 00 (00) | vv    (vv)  | Bank Select (MSB) vv=0                      *1-1    |
...
|   Bn   | 20 (32) | vv    (vv)  | Bank Select (LSB) vv=0~4                    *1-1    |
...
 *1-1 : This message is transmitted when the "MIDI Tx Prog Chg" is set to "On".
,,,

 2.RECOGNIZED RECEIVE DATA

2-1 CHANNEL MESSAGES
+--------+---------+-------------+--------------------------------------------------+
| Status | Second  |   Third     |           Description                            |
| [Hex]  | [H] [D] | [H]   [D]   |                                                  |
+--------+---------+-------------+--------------------------------------------------+
...
|   9n   | kk (kk) | vv    (vv)  | Note On  vv=1~127                                |
|   Bn   | 00 (00) | vv    (vv)  | Bank Select (MSB) vv=0                   *5-1    |
...
|   Bn   | 20 (32) | vv    (vv)  | Bank Select (LSB) vv=0~4                 *5-1    |
...
 *5-1 : This message is recognized when the "MIDI Rx Prog Chg" is set to "On".
```

Since there are only 5 banks, the Bank MSB value is ignored and doesn't need to be sent, based on experience.   Patches are selected from the panel as 1-500, but via MIDI, they're in 5 banks of 100.

IIRC, The Mutable Instruments Shruthi=1 does it too, but there are different firmware versions in use, some added to by contributors, so I'm not sure if that applies to all of them.

A lot of synths only have a few banks, sometimes with the maximum 128 patches, so I wouldn't be surprised if it's fairly common.  I see it as being like "running status", where you're not required to send the status byte again, if it hasn't changed, because it's remembered in the state information of the synth.  

Obviously, using a merger could complicate that so a merger needs to allow for it, or just expand each message by resending the status byte if it's missing.  If it didn't allow for bank selects coming from different inputs, that would cause problems, but it's not likely to be a problem in practice - there's likely to be one MIDI controller connected that has the job of sending bank selects and program changes, while others send CCs and Note On/Off.

At the moment, the BANKSEL MSB seems to behave as though it's remembered as part of the state, because I can just send BANKSEL LSB, then Program Change, and it switches to the new program in the new bank.  If changes to MSB could be done in the same way, it would be simpler to set up a knob for each, or maybe increment and decrement buttons, if the controller allows it.  

It could be done in Ctrlr, for example, running on a PC, optionally with physical knobs from a simple MIDI controller assigned to work the on-screen knobs - e.g. I can do that in Cantabile (a DAW) using Ctrlr as a VST.  Knobs sending a range of 0-127 are fairly practical, especially if they're clicky encoders, but a knob sending 0-16383 isn't, of course.

The MIDI spec seems to describe expected behaviour, cautioning that a controller sending anything else, or in a different way, might not work as you expect.  I don't read that as saying a synth can't be allowed to cope with it done differently, just that you shouldn't expect it to work necessarily, if you do.

**Comment by poetaster** [comment]
> Hi poetaster,
> 
> > I have three folders in sysex/voices folder (oo1blah, 002foo, 003bar) and can't keep track. What would be a could test strategy? Or should I simply be naming the files with bank info? The docs are a bit confusing. And I'm a bit confused :)
> 
> The file names of the syx files are crucial. Look at the following folder structure with the sysex files: MiniDexed-sysexFinal.zip [#456 (comment)](https://github.com/probonopd/MiniDexed/discussions/456#discussioncomment-6406179)
> 
> Peter 🍌

Yes, that's all clear. The question is how do PROG MSB LSB hang together? I was hoping some numeral indexing would work, but, not really. I'm just going to test with 3 folders with 3 voices each and see. I was thinking it's be, LSB folder, MSB voice and Program, program.
